src/hypervisor_bootloader/src/dtb.rs

```diff
@@ -602,3 +602,111 @@ impl DtbAnalyser {
         self.struct_block_size + self.struct_block_address
     }
 }
+
+#[cfg(feature = "raspberrypi")]
+pub fn add_new_reserved_area_to_dtb(
+    original_base_address: usize,
+    new_base_address: usize,
+    new_size: usize,
+    reserved_address: usize,
+    reserved_size: usize,
+) -> Result<(), ()> {
+    let mut total_new_size = 0;
+    let original_dtb_header = unsafe { &*(original_base_address as *const DtbHeader) };
+
+    let new_dtb_header = unsafe { &mut *(new_base_address as *mut DtbHeader) };
+    total_new_size += core::mem::size_of::<DtbHeader>();
+    if new_size < total_new_size {
+        return Err(());
+    }
+
+    new_dtb_header.magic = original_dtb_header.magic;
+    new_dtb_header.version = original_dtb_header.version;
+    new_dtb_header.last_comp_version = original_dtb_header.last_comp_version;
+    new_dtb_header.boot_cpuid_phys = original_dtb_header.boot_cpuid_phys;
+    new_dtb_header.size_dt_struct = original_dtb_header.size_dt_struct;
+    new_dtb_header.size_dt_strings = original_dtb_header.size_dt_strings;
+
+    // copy reserved address section and add new reserved section
+    let original_reserved_address_section =
+        original_base_address + u32::from_be(original_dtb_header.off_mem_rsv_map) as usize;
+    let new_reserved_address_section = new_base_address + total_new_size;
+    let mut pointer = original_reserved_address_section;
+    loop {
+        let address = unsafe { *(pointer as *const u64) };
+        pointer += core::mem::size_of::<u64>();
+        let size = unsafe { *(pointer as *const u64) };
+        pointer += core::mem::size_of::<u64>();
+        if address == 0 && size == 0 {
```

Comment:
If you increase the counter (`pointer`) after this check, you do not need to decrease `core::mem::size_of::<u64>() * 2` at Line 645.

Reply:
No. I want to overwrite the last node with new reservation block address and size, and then add last termination entry. So, that line is necessary.